source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,259 @@
+#include "common/network/apple_dns_impl.h"
+
+#include <dns_sd.h>
+
+#include <chrono>
+#include <cstdint>
+#include <list>
+#include <memory>
+#include <string>
+
+#include "envoy/common/platform.h"
+
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+#include "common/network/address_impl.h"
+#include "common/network/utility.h"
+
+#include "absl/strings/str_join.h"
+
+namespace Envoy {
+namespace Network {
+
+AppleDnsResolverImpl::AppleDnsResolverImpl(Event::Dispatcher& dispatcher)
+    : dispatcher_(dispatcher) {
+  ENVOY_LOG(warn, "Constructing DNS resolver");
+  initializeMainSdRef();
+}
+
+AppleDnsResolverImpl::~AppleDnsResolverImpl() {
+  ENVOY_LOG(warn, "Destructing DNS resolver");
+  deallocateMainSdRef();
+}
+
+void AppleDnsResolverImpl::deallocateMainSdRef() {
+  ENVOY_LOG(warn, "DNSServiceRefDeallocate main sd ref");
+  // dns_sd.h says:
+  //   If the reference's underlying socket is used in a run loop or select() call, it should
+  //   be removed BEFORE DNSServiceRefDeallocate() is called, as this function closes the
+  //   reference's socket.
+  sd_ref_event_.reset();
+  DNSServiceRefDeallocate(main_sd_ref_);
+}
+
+void AppleDnsResolverImpl::initializeMainSdRef() {
+  auto error = DNSServiceCreateConnection(&main_sd_ref_);
+  if (error) {
+    throw EnvoyException("Error in DNSServiceCreateConnection");
+  }
+
+  auto fd = DNSServiceRefSockFD(main_sd_ref_);
+  if (fd == -1) {
+    throw EnvoyException("Error in DNSServiceRefSockFD");
+  }
+  ENVOY_LOG(warn, "DNS resolver has fd={}", fd);
+
+  sd_ref_event_ = dispatcher_.createFileEvent(
+      fd, [this](uint32_t events) { onEventCallback(events); }, Event::FileTriggerType::Level,
```

Comment:
Do you for sure want level triggering here? It's probably fine as long as you always drain all events, but I might add a comment as to why you selected level triggering.

Reply:
I have to be honest, here I ended up doing level triggering to more closely resemble the c-ares based implementation.

Reply:
OK I would just drop a comment. I think level should be fine as you will consume all events.